pip/pip-385.md

```diff
@@ -0,0 +1,348 @@
+maxMessagePublishBufferSizeInMB or broker level rate limit - **we continue to pause the connection**, but we still send
+the `ThrottleProducer` command in order to inform the client about the reason for any potential timeout. The reason we
+continue to pause reads is that we are already breaching memory limits, thus, even if the client sends
+a `ThrottleProducerReceipt` response, we won't be able to read it until further pending messages before that are read.
+
+Here is a sequence diagram highlighting the case when a topic level breach happens:
+
+```mermaid
+%%{init: {"mirrorActors": false, "rightAngles": false} }%%
+sequenceDiagram
+    Client->>Server: CreateProducer(reqId, myTopic)
+    Note right of Server: Check Authorization
+    Server-->>Client: ProducerSuccess(reqId, producerName)
+    Activate Client
+    Activate Server
+    Client->>Server: Send(1, message1)
+    Client->>Server: Send(2, message2)
+    Server-->>Client: SendReceipt(1, msgId1)
+    Note right of Server: Check Throttling
+
+    opt topic/connection throttled
+        Server->>Client: ThrottleProducer(reqId, throttleData)
+        alt client sends receipt in time
+            Client-->>Server: ThrottleProducerReceipt(reqId)
+            note right of Client: client pauses till specified time
+            Note over Client, Server: After some time
+            Client->>Server: Send(3, message3)
+            Client->>Server: Send(4, message4)
+
+        else no response
+            Activate Server
+            note right of Server: TCP read paused
+            Client-xServer: Send(3, message3)
+            Note over Client, Server: After some time
+            opt topic/connection unthrottled
+                note right of Server: TCP read resumed
+                deactivate Server
+                Server-->>Server: read message 3
+                Client->>Server: Send(4, message4)
+            end
+        end
+    end
+    Server-->>Client: SendReceipt(2,msgId2)
+    Server-->>Client: SendReceipt(3,msgId3)
+    Server-->>Client: SendReceipt(4,msgId4)
+
+    Client->>Server: CloseProducer(producerId, reqId)
+    Server-->>-Client: Success(reqId)
+    deactivate Client
+```
+
+## Java client changes
+
+* Client will now have logic to understand the `ThrottleProducer` command and take relevant action of blocking further
+  messages for the relevant topic. It will then respond back with `ThrottleProducerReceipt` command.
+    * Client will resume message sending after the specified time in the `ThrottleProducer` command's data.
+    * This interval of no messages will be noted as "being throttled"
+    * Within this duration, another `ThrottleProducer` command from server may come.
+* Producer will record new OTel metric indicating which topic was throttled and the reason.
+* In case a message fails due to timeout and there was a throttled command from server for the owning topic, client will
+  instead throw a rate limit exception instead of timeout exception.
```

Comment:
That's a good point! For a sendTimeout of 2 seconds its possible that most of the time after the message was sent, the producer was being throttled, but just not right before the 2 second mark.

The `T` in the equation should be the `now() - createAt` and the `B` should also be calculated for the same time period. The calculation doesn't have to be precise in favor of being fast.
I suppose we don't want to give the threshold lever out to clients as this is not really something that people would be worrying about (when they are getting timeout vs rate limit exception) too much.

Reply:
@lhotari  - I've added a section explaining the same.

Reply:
@lhotari  - Hi Lari, if you don't have any further comments, can you resolve this one?

Reply:
@grssam If possible, we use the timeout exception to wrap the limit exception. Sometimes, users want to know why the timeout occurs on the client side.

Reply:
@nodece  - This will add a new client exception all together. Technically, we will be wrapping the PulsarClientException to the limit exception.
At that point, is there any specific need to still put in the time out exception? It would be implicit in my opinion.